src/interrupt/mod.rs

```diff
@@ -0,0 +1,369 @@
+//!
+//! A device may support multiple types of interrupts, and each type of interrupt may support one
+//! or multiple interrupt sources. For example, a PCI device may support:
+//! * Legacy Irq: exactly one interrupt source.
+//! * PCI MSI Irq: 1,2,4,8,16,32 interrupt sources.
+//! * PCI MSIx Irq: 2^n(n=0-11) interrupt sources.
+//!
+//! A distinct Interrupt Source Identifier (ISID) will be assigned to each interrupt source.
+//! An ID allocator will be used to allocate and free Interrupt Source Identifiers for devices.
+//! To decouple the vm-device crate from the ID allocator, the vm-device crate doesn't take the
+//! responsibility to allocate/free Interrupt Source IDs but only makes use of assigned IDs.
+
+use std::fmt::{self, Display};
+use std::io;
+use std::ops::Deref;
+use std::ops::Index;
+use vmm_sys_util::eventfd::EventFd;
+
+/// Errors associated with handling interrupts
+#[derive(Debug)]
+pub enum Error {
+    /// Operation not supported for this interrupt.
+    OperationNotSupported,
+
+    /// The specified configuration is not valid.
+    InvalidConfiguration,
+
+    /// The interrupt was not enabled.
+    InterruptNotEnabled,
+
+    /// Generic IO error,
+    IOError(io::Error),
+}
+
+/// Reuse std::io::Result to simplify interoperability among crates.
+pub type Result<T> = std::result::Result<T, Error>;
+
+impl std::error::Error for Error {}
+
+impl Display for Error {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "Interrupt error: ")?;
+        match self {
+            Error::OperationNotSupported => write!(f, "operation not supported"),
+            Error::InvalidConfiguration => write!(f, "invalid configuration"),
+            Error::InterruptNotEnabled => write!(f, "the interrupt was not enabled"),
+            Error::IOError(error) => write!(f, "{}", error),
+        }
+    }
+}
+
+/// Data type to store an interrupt source identifier.
+pub type InterruptIndex = u32;
+
+pub trait Interrupt {
+    /// Type of configuration information for interrupt source.
+    type C;
+
+    /// Enable generation of interrupts from this interrupt source.
+    fn enable(&self, config: &Self::C) -> Result<()> {
+        self.update(config)
+    }
+
+    /// Disable generation of interrupts from this interrupt source.
+    fn disable(&self) -> Result<()> {
+        Err(Error::OperationNotSupported)
+    }
+
+    /// Update configuration of the interrupt.
+    fn update(&self, _config: &Self::C) -> Result<()> {
+        Err(Error::OperationNotSupported)
+    }
+
+    /// Returns an interrupt notifier from this interrupt.
+    ///
+    /// An interrupt notifier allows for external components and processes
+    /// to inject interrupts into a guest, by writing to the file returned
+    /// by this method.
+    fn notifier(&self) -> Option<&EventFd> {
+        None
+    }
+
+    /// Inject an interrupt from this interrupt source into the guest.
+    fn trigger(&self) -> Result<()>;
```

Comment:
For level triggered interrupts, do we need a corresponding stop_triggering() function?

Reply:
I would rather avoid making the interrupt trait stateful. Instead, I would add an `eoi()` method instead. Level-triggered interrupts can check if the level is stil 1 and if so re-inject. The method would be invoked logically from the CPU, but it would also map nicely to KVM resamplefd semantics so there should be another `resample_notifier()` method.